File: ios/Sources/UnifiedTrackingPlugin/Analytics/AmplitudeProvider.swift
import Foundation
import AmplitudeSwift

/// Amplitude (Amplitude-iOS / amplitude-swift) bridge for the unified-tracking plugin.
///
/// Reads `apiKey` (required) and optional `serverUrl` from the provider config and
/// initializes an Amplitude instance. If no API key is supplied the SDK is not
/// started and the provider stays uninitialized.
///
/// NOTE(unverified): build with Xcode — confirm the module import name. The newer
/// Swift-native SDK ships as `AmplitudeSwift` while the legacy Objective-C SDK
/// ships as `Amplitude` (`import Amplitude`). The `Package.swift` references the
/// `Amplitude-iOS` package which historically vended the ObjC SDK, but the
/// product name in current versions is typically `AmplitudeSwift`. Verify which
/// product the resolved package version exposes and adjust the import + API.
class AmplitudeProvider: AnalyticsProvider {
    let name = "amplitude"
    private(set) var isEnabled: Bool
    private(set) var isInitialized: Bool

    private let config: [String: Any]
    private var debugMode = false
    /// Tracks the consent state for parity. Amplitude exposes `setOptOut(...)`.
    private var consentGranted = true
    /// The Amplitude instance returned by `Amplitude(configuration:)`.
    private var instance: Amplitude?

    init(config: [String: Any]) {
        self.config = config
        self.isEnabled = config["enabled"] as? Bool ?? true
        self.isInitialized = false

        guard let apiKey = config["apiKey"] as? String, !apiKey.isEmpty else {
            // No API key -> cannot start Amplitude. Leave uninitialized.
            print("[AmplitudeProvider] No apiKey provided; Amplitude not started.")
            return
        }

        // NOTE(unverified): build with Xcode — confirm `Configuration(apiKey:serverUrl:)`
        // initializer and the `Amplitude(configuration:)` initializer signatures in
        // amplitude-swift / AmplitudeSwift ~1.x. The legacy ObjC SDK uses
        // `Amplitude.instance().initializeApiKey(apiKey)` instead — adjust if the
        // resolved package exposes the legacy API.
        let configuration: Configuration
        if let serverUrl = config["serverUrl"] as? String, !serverUrl.isEmpty {
            configuration = Configuration(apiKey: apiKey, serverUrl: serverUrl)
        } else {
            configuration = Configuration(apiKey: apiKey)
        }
        self.instance = Amplitude(configuration: configuration)

        self.isInitialized = true
    }

    func trackEvent(_ event: String, properties: [String: Any]) {
        guard isEnabled && isInitialized && consentGranted, let instance = instance else { return }
        // NOTE(unverified): build with Xcode — confirm `track(eventType:eventProperties:)`
        // argument labels in AmplitudeSwift ~1.x. Legacy ObjC SDK uses
        // `logEvent(_:withEventProperties:)` instead.
        instance.track(eventType: event, eventProperties: properties)
    }

    func identifyUser(_ userId: String, traits: [String: Any]) {
        guard isEnabled && isInitialized && consentGranted, let instance = instance else { return }
        // NOTE(unverified): build with Xcode — confirm `setUserId(userId:)` argument label
        // (some SDK versions expose `setUserId(_:)` without a label).
        instance.setUserId(userId: userId)
        if !traits.isEmpty {
            // NOTE(unverified): build with Xcode — confirm `Identify()` initializer and
            // `set(property:value:)` chainable signature in AmplitudeSwift ~1.x.
            let identify = Identify()
            for (key, value) in traits {
                identify.set(property: key, value: value)
            }
            instance.identify(identify: identify)
        }
    }

    func setUserProperties(_ properties: [String: Any]) {
        guard isEnabled && isInitialized && consentGranted, let instance = instance else { return }
        let identify = Identify()
        for (key, value) in properties {
            identify.set(property: key, value: value)
        }
        instance.identify(identify: identify)
    }

    func logRevenue(amount: Double, currency: String, productId: String?, quantity: Int, properties: [String: Any]) {
        guard isEnabled && isInitialized && consentGranted, let instance = instance else { return }
        // NOTE(unverified): build with Xcode — confirm `Revenue()` initializer and the
        // chainable `setProductId(_:)` / `setQuantity(_:)` / `setPrice(_:)` setters in
        // AmplitudeSwift ~1.x. Some SDK versions expose these as plain property setters
        // (e.g. `revenue.productId = ...`) instead of chainable methods.
        let revenue = Revenue()
        if let productId = productId {
            revenue.setProductId(productId)
        }
        revenue.setQuantity(quantity)
        revenue.setPrice(amount)
        // Currency + extra properties are not first-class on the Revenue model in
        // every SDK version. We attach them via `eventProperties` when supported.
        // NOTE(unverified): build with Xcode — some versions expose
        // `revenue.eventProperties = ...` or `revenue.setEventProperties(_:)`.
        var extras: [String: Any] = ["currency": currency]
        for (key, value) in properties {
            extras[key] = value
        }
        revenue.setEventProperties(extras)
        instance.revenue(revenue: revenue)
    }

    func logScreenView(_ screenName: String, properties: [String: Any]) {
        guard isEnabled && isInitialized && consentGranted, let instance = instance else { return }

        var params: [String: Any] = [
            // Amplitude's canonical screen-view property names.
            "[Amplitude] Screen Name": screenName
        ]
        for (key, value) in properties {
            params[key] = value
        }

        // Amplitude's canonical screen-view event name.
        instance.track(eventType: "[Amplitude] Screen Viewed", eventProperties: params)
    }

    func setConsent(_ granted: Bool) {
        consentGranted = granted
        guard let instance = instance else { return }
        // NOTE(unverified): build with Xcode — confirm `setOptOut(_:)` exists as an
        // instance method in AmplitudeSwift ~1.x. The legacy ObjC SDK exposes
        // `Amplitude.instance().optOut = true` instead.
        instance.setOptOut(optOut: !granted)
    }

    func reset() {
        guard let instance = instance else { return }
        // NOTE(unverified): build with Xcode — confirm `reset()` is an instance method
        // on `Amplitude` in AmplitudeSwift ~1.x. Legacy ObjC SDK exposes `regenerateDeviceId()`.
        instance.reset()
    }

    func setDebugMode(_ enabled: Bool) {
        // Amplitude's log level is configured at SDK init time via the `Configuration`
        // object's `logLevel` property. We retain the flag here for parity; flipping
        // it at runtime is not exposed as a first-class API in all SDK versions.
        debugMode = enabled
    }
}

File: android/src/main/java/com/aoneahsan/unifiedtracking/providers/analytics/AmplitudeProvider.java
package com.aoneahsan.unifiedtracking.providers.analytics;

import android.content.Context;
import android.util.Log;

import com.getcapacitor.JSObject;
import com.aoneahsan.unifiedtracking.providers.AnalyticsProvider;
import com.amplitude.android.Amplitude;
import com.amplitude.android.Configuration;
import com.amplitude.core.events.Identify;
import com.amplitude.core.events.Revenue;

import java.util.HashMap;
import java.util.Map;

/**
 * Amplitude analytics provider for Android.
 *
 * Uses {@code com.amplitude.android.Amplitude} (the Kotlin-based Amplitude Android SDK,
 * artifact {@code com.amplitude:analytics-android}). Java consumers instantiate the same
 * classes — {@code Amplitude} + {@code Configuration} — exposed via Kotlin interop.
 *
 * NOTE(unverified): built against Amplitude Android SDK 1.x (artifact
 * {@code com.amplitude:analytics-android:1.x}). The dependency declares
 * {@code amplitudeVersion = '3.48.2'} which historically referenced the legacy
 * {@code com.amplitude:android-sdk:2.x} package; the current 1.x analytics-android
 * series replaced it. Confirm in Android Studio that the resolved artifact exposes
 * {@code com.amplitude.android.Amplitude} and {@code com.amplitude.android.Configuration};
 * if the project is still on the legacy SDK the API surface differs.
 */
public class AmplitudeProvider implements AnalyticsProvider {
    private static final String TAG = "Amplitude";
    private final Context context;
    private final JSObject config;
    private Amplitude amplitude;
    private boolean enabled;
    private boolean initialized;
    private boolean debugMode = false;

    public AmplitudeProvider(Context context, JSObject config) {
        this.context = context;
        this.config = config;
        this.enabled = config.getBoolean("enabled", true);
        this.initialized = false;

        try {
            String apiKey = config.getString("apiKey");
            if (apiKey == null || apiKey.trim().isEmpty()) {
                Log.e(TAG, "Amplitude apiKey is missing; provider will be inert.");
                return;
            }

            // NOTE(unverified): Configuration(String, Context) is the documented 2-arg
            // primary constructor in Amplitude Android SDK 1.x; additional options (
            // serverZone, flushQueueSize, etc.) can be set via setters after construction.
            // Confirm exact constructor + property setters against the resolved SDK.
            Configuration cfg = new Configuration(apiKey, context);
            this.amplitude = new Amplitude(cfg);
            this.initialized = true;

            if (debugMode) {
                Log.d(TAG, "Amplitude initialized");
            }
        } catch (Exception e) {
            Log.e(TAG, "Failed to initialize Amplitude", e);
            this.initialized = false;
        }
    }

    @Override
    public String getName() {
        return "amplitude";
    }

    @Override
    public boolean isEnabled() {
        return enabled;
    }

    @Override
    public boolean isInitialized() {
        return initialized;
    }

    @Override
    public void trackEvent(String event, Map<String, Object> properties) {
        if (!enabled || !initialized || amplitude == null || event == null) return;

        try {
            // NOTE(unverified): Amplitude.track(String, Map<String, Object>) is the
            // documented overload in Amplitude Android SDK 1.x. Confirm the second
            // argument accepts Map<String, Object> vs requiring a JSONObject.
            if (properties != null) {
                amplitude.track(event, properties);
            } else {
                amplitude.track(event);
            }

            if (debugMode) {
                Log.d(TAG, "Tracked event: " + event);
            }
        } catch (Exception e) {
            Log.e(TAG, "Failed to track event", e);
        }
    }

    @Override
    public void identifyUser(String userId, Map<String, Object> traits) {
        if (!enabled || !initialized || amplitude == null || userId == null) return;

        try {
            amplitude.setUserId(userId);

            if (traits != null && !traits.isEmpty()) {
                // NOTE(unverified): com.amplitude.core.events.Identify provides set(String, Object)
                // for user-property updates in Amplitude SDK 1.x. Identify is applied via
                // Amplitude.identify(Identify) which queues a $identify event.
                Identify identify = new Identify();
                for (Map.Entry<String, Object> entry : traits.entrySet()) {
                    if (entry.getKey() == null || entry.getValue() == null) continue;
                    identify.set(entry.getKey(), entry.getValue());
                }
                amplitude.identify(identify);
            }

            if (debugMode) {
                Log.d(TAG, "Identified user: " + userId);
            }
        } catch (Exception e) {
            Log.e(TAG, "Failed to identify user", e);
        }
    }

    @Override
    public void setUserProperties(Map<String, Object> properties) {
        if (!enabled || !initialized || amplitude == null || properties == null) return;

        try {
            Identify identify = new Identify();
            for (Map.Entry<String, Object> entry : properties.entrySet()) {
                if (entry.getKey() == null || entry.getValue() == null) continue;
                identify.set(entry.getKey(), entry.getValue());
            }
            amplitude.identify(identify);

            if (debugMode) {
                Log.d(TAG, "Set user properties");
            }
        } catch (Exception e) {
            Log.e(TAG, "Failed to set user properties", e);
        }
    }

    @Override
    public void logRevenue(double amount, String currency, String productId, int quantity, Map<String, Object> properties) {
        if (!enabled || !initialized || amplitude == null) return;

        try {
            // NOTE(unverified): com.amplitude.core.events.Revenue exposes fluent setters
            // (setProductId/setQuantity/setPrice) in Amplitude SDK 1.x. The revenue is
            // submitted via Amplitude.revenue(Revenue). Earlier SDKs used a builder
            // pattern; confirm against the resolved Revenue class.
            Revenue revenue = new Revenue();
            if (productId != null) revenue.setProductId(productId);
            revenue.setQuantity(quantity);
            revenue.setPrice(amount);

            if (properties != null && !properties.isEmpty()) {
                // Currency + extra properties are attached via Revenue.setEventProperties
                // as a JSON-like map on most SDK versions.
                Map<String, Object> revenueProps = new HashMap<>(properties);
                if (currency != null && !revenueProps.containsKey("currency")) {
                    revenueProps.put("currency", currency);
                }
                try {
                    // NOTE(unverified): Revenue.setEventProperties(Map<String,Object>) is
                    // typically present; if absent on the resolved SDK, fall back to
                    // tracking the currency on a separate $revenue event property.
                    revenue.setEventProperties(new org.json.JSONObject(revenueProps));
                } catch (Throwable ignored) {
                    // Silently skip if the setter is unavailable on the resolved SDK.
                }
            } else if (currency != null) {
                try {
                    org.json.JSONObject extra = new org.json.JSONObject();
                    extra.put("currency", currency);
                    revenue.setEventProperties(extra);
                } catch (Throwable ignored) {
                    // Optional best-effort attachment.
                }
            }

            amplitude.revenue(revenue);

            if (debugMode) {
                Log.d(TAG, "Logged revenue: " + amount + " " + currency);
            }
        } catch (Exception e) {
            Log.e(TAG, "Failed to log revenue", e);
        }
    }

    @Override
    public void logScreenView(String screenName, Map<String, Object> properties) {
        if (!enabled || !initialized || amplitude == null || screenName == null) return;

        try {
            Map<String, Object> props = new HashMap<>();
            if (properties != null) {
                props.putAll(properties);
            }
            // Use Amplitude's canonical screen-view convention so dashboards pick it up.
            props.put("[Amplitude] Screen Name", screenName);
            amplitude.track("[Amplitude] Screen Viewed", props);

            if (debugMode) {
                Log.d(TAG, "Logged screen view: " + screenName);
            }
        } catch (Exception e) {
            Log.e(TAG, "Failed to log screen view", e);
        }
    }

    @Override
    public void setConsent(boolean granted) {
        if (!initialized || amplitude == null) return;

        try {
            // NOTE(unverified): Amplitude SDK 1.x exposes setOptOut(boolean). When
            // optOut=true the SDK drops events locally; confirm the property name on
            // the resolved Configuration/Amplitude class (some SDK builds expose it
            // via amplitude.getConfiguration().setOptOut(...) instead).
            amplitude.setOptOut(!granted);
            this.enabled = granted;

            if (debugMode) {
                Log.d(TAG, "Set consent: " + granted);
            }
        } catch (Exception e) {
            Log.e(TAG, "Failed to set consent", e);
        }
    }

    @Override
    public void reset() {
        if (!initialized || amplitude == null) return;

        try {
            // NOTE(unverified): Amplitude.reset() clears the user/device identity in
            // SDK 1.x. Older SDKs used setUserId(null) + regenerateDeviceId(); confirm
            // reset() is present on the resolved Amplitude class.
            amplitude.reset();

            if (debugMode) {
                Log.d(TAG, "Reset Amplitude state");
            }
        } catch (Exception e) {
            Log.e(TAG, "Failed to reset", e);
        }
    }

    @Override
    public void setDebugMode(boolean enabled) {
        this.debugMode = enabled;
        if (enabled) {
            Log.d(TAG, "Debug mode enabled for Amplitude provider");
        }
    }
}
